File: lib/krypto/challenge.rb
# frozen_string_literal: true

require "rbnacl"
require "msgpack"
require "openssl"

module Krypto

  MAX_CHALLENGE_SIZE = 4 * 1024 * 1024

  class Challenge
    def self.generate(signing_key, challenge_id, challenge_data, request_data, timestamp: Time.now)
      private_encryption_key = RbNaCl::PrivateKey.generate
      public_encryption_key = private_encryption_key.public_key

      msg = MessagePack.pack(
        InnerChallenge.new(
          publicEncryptionKey: public_encryption_key.to_bytes,
          challengeData: challenge_data,
          requestData: request_data,
          timestamp: timestamp.to_i,
          challengeId: challenge_id
        )
      )

      outer = OuterChallenge.new(
        sig: signing_key.sign(OpenSSL::Digest.new("SHA256"), msg),
        msg: msg
      )

      [outer, private_encryption_key.to_bytes]
    end

    def self.unmarshal(data, png: false, base64: true)
      if data.size > MAX_CHALLENGE_SIZE
        raise "challenge too large"
      end

      data = ::Krypto::Png.decode_blob(data) if png
      data = Base64.strict_decode64(data) if base64
      OuterChallenge.new(MessagePack.unpack(data).slice(*OUTER_CHALLENGE_FIELDS.map(&:to_s)))
    end

    # OuterChallenge is meant as the main challenge entry point. It is a simple wrapper over an internal message.
    OUTER_CHALLENGE_FIELDS = %i[sig msg].freeze
    class OuterChallenge < Struct.new(*OUTER_CHALLENGE_FIELDS, keyword_init: true)
      def to_msgpack(out = "")
        to_h.to_msgpack(out)
      end

      def verify(key)
        return false unless Krypto::Ec.verify(key, sig, msg)

        @inner = InnerChallenge.new(MessagePack.unpack(msg))

        true
      end

      def timestamp
        @inner&.timestamp
      end

      def request_data
        @inner&.requestData
      end

      # respond creates a reponse to the challenge. It accepts 2 keys to sign
      # the response with, the second may be nil.
      def respond(signing_key, signing_key_2, response_data)
        raise "No inner. Unverified?" unless @inner

        public_signing_key_2_der = if !signing_key_2.nil?
          Base64.strict_encode64(signing_key_2.public_to_der)
        else
          ""
        end

        msg = MessagePack.pack(
          Krypto::ChallengeResponse::InnerResponse.new(
            publicSigningKey: Base64.strict_encode64(signing_key.public_to_der),
            publicSigningKey2: public_signing_key_2_der,
            challengeData: @inner.challengeData,
            responseData: response_data,
            timestamp: Time.now.to_i
          )
        )

        sig = Krypto::Ec.sign(signing_key, msg)
        private_encryption_key = RbNaCl::PrivateKey.generate

        sig2 = if !signing_key_2.nil?
          Krypto::Ec.sign(signing_key_2, msg)
        else
          ""
        end

        box = RbNaCl::SimpleBox.from_keypair(@inner.publicEncryptionKey, private_encryption_key)
        sealed = box.encrypt(msg)

        Krypto::ChallengeResponse::OuterResponse.new(
          sig: sig,
          sig2: sig2,
          publicEncryptionKey: private_encryption_key.public_key.to_bytes,
          msg: sealed,
          challengeId: @inner.challengeId
        )
      end
    end

    INNER_CHALLENGE_FIELDS = %i[publicEncryptionKey challengeData requestData timestamp challengeId].freeze
    class InnerChallenge < Struct.new(*INNER_CHALLENGE_FIELDS, keyword_init: true)
      def to_msgpack(out = "")
        to_h.to_msgpack(out)
      end
    end
  end
end

File: pkg/challenge/challenge.go
package challenge

import (
	"bytes"
	"crypto"
	"crypto/ecdsa"
	"crypto/rand"
	"fmt"
	"time"

	"github.com/kolide/krypto"
	"github.com/kolide/krypto/pkg/echelper"
	"github.com/vmihailenco/msgpack/v5"
	"golang.org/x/crypto/nacl/box"
)

const (
	signingTimeoutDuration = 1 * time.Second
	signingTimeoutInterval = 250 * time.Millisecond
)

type OuterChallenge struct {
	Sig            []byte `msgpack:"sig"`
	Msg            []byte `msgpack:"msg"`
	innerChallenge *InnerChallenge
}

func (o *OuterChallenge) Verify(counterParty ecdsa.PublicKey) error {
	if err := echelper.VerifySignature(&counterParty, o.Msg, o.Sig); err != nil {
		return err
	}

	innerChallenge, err := o.inner()
	if err != nil {
		return err
	}

	o.innerChallenge = innerChallenge
	return nil
}

func (o *OuterChallenge) RequestData() []byte {
	if o.innerChallenge != nil {
		return o.innerChallenge.RequestData
	}

	return nil
}

func (o *OuterChallenge) Timestamp() int64 {
	if o.innerChallenge != nil {
		return o.innerChallenge.Timestamp
	}

	return -1
}

func (o *OuterChallenge) Marshal() ([]byte, error) {
	return msgpack.Marshal(o)
}

// Respond creates a response to the challenge. It accepts crypto.Signer for signing, the second one may be nil.
// If the second signer is not nil and produces an error while signing, it will not be included nor will an error
// be returned. This is because the second signer is intended to be a TPM or Secure Enclave which we have found to be
// unreliable in testing.
func (o *OuterChallenge) Respond(signer crypto.Signer, signer2 crypto.Signer, responseData []byte) ([]byte, error) {
	if o.innerChallenge == nil {
		return nil, fmt.Errorf("no inner. unverified?")
	}

	ecdsaPubKey, ok := signer.Public().(*ecdsa.PublicKey)
	if !ok {
		return nil, fmt.Errorf("signer pubkey in unexpected format (expected ECDSA, got %T)", signer.Public())
	}

	pubSigningDer, err := echelper.PublicEcdsaToB64Der(ecdsaPubKey)
	if err != nil {
		return nil, fmt.Errorf("marshalling public signing key to der: %w", err)
	}

	var pubSigningKey2Der []byte
	if signer2 != nil {
		ecdsaPubKey2, ok := signer2.Public().(*ecdsa.PublicKey)
		if !ok {
			return nil, fmt.Errorf("signer pubkey 2 in unexpected format (expected ECDSA, got %T)", signer2.Public())
		}

		pubSigningKey2Der, err = echelper.PublicEcdsaToB64Der(ecdsaPubKey2)
		if err != nil {
			return nil, fmt.Errorf("marshalling public signing 2 key to der: %w", err)
		}
	}

	innerResponse, err := msgpack.Marshal(InnerResponse{
		PublicSigningKey:  pubSigningDer,
		PublicSigningKey2: pubSigningKey2Der,
		ChallengeData:     o.innerChallenge.ChallengeData,
		ResponseData:      responseData,
		Timestamp:         time.Now().Unix(),
	})

	if err != nil {
		return nil, fmt.Errorf("marshaling inner box: %w", err)
	}

	signature, err := echelper.SignWithTimeout(signer, innerResponse, signingTimeoutDuration, signingTimeoutInterval)
	if err != nil {
		return nil, fmt.Errorf("signing challenge: %w", err)
	}

	var signature2 []byte
	if signer2 != nil {
		// Intentionally ignoring errors here because we use the second signature to sign with a TPM or Secure Enclave,
		// we have found both of these to be unreliable in testing and we are not using them in any meaningful way.
		//nolint: errcheck - we allow nil signer2
		signature2, _ = echelper.SignWithTimeout(signer2, innerResponse, signingTimeoutDuration, signingTimeoutInterval)
	}

	sealed, pub, err := echelper.SealNaCl(innerResponse, &o.innerChallenge.PublicEncryptionKey)
	if err != nil {
		return nil, fmt.Errorf("sealing challenge inner box: %w", err)
	}

	outerResponse := &OuterResponse{
		PublicEncryptionKey: *pub,
		Sig:                 signature,
		Sig2:                signature2,
		Msg:                 sealed,
		ChallengeId:         o.innerChallenge.ChallengeId,
	}

	return msgpack.Marshal(outerResponse)
}

func (o *OuterChallenge) inner() (*InnerChallenge, error) {
	if o.Msg == nil {
		return nil, fmt.Errorf("inner is nil")
	}

	var inner InnerChallenge
	if err := msgpack.Unmarshal(o.Msg, &inner); err != nil {
		return nil, fmt.Errorf("unmarshaling inner challenge: %w", err)
	}
	return &inner, nil
}

func (o *OuterChallenge) RespondPng(signer crypto.Signer, signer2 crypto.Signer, responseData []byte) ([]byte, error) {
	response, err := o.Respond(signer, signer2, responseData)
	if err != nil {
		return nil, err
	}

	var buf bytes.Buffer
	if err := krypto.ToPng(&buf, response); err != nil {
		return nil, fmt.Errorf("encoding data to png: %w", err)
	}

	return buf.Bytes(), nil
}

func UnmarshalChallenge(outerChallengeBytes []byte) (*OuterChallenge, error) {
	if len(outerChallengeBytes) > krypto.V0MaxSize {
		return nil, fmt.Errorf("challenge exceeds max size: %d, max is %d", len(outerChallengeBytes), krypto.V0MaxSize)
	}

	var outerChallenge OuterChallenge
	if err := msgpack.Unmarshal(outerChallengeBytes, &outerChallenge); err != nil {
		return nil, err
	}
	return &outerChallenge, nil
}

type InnerChallenge struct {
	// PublicEncryptionKey is the public half of the NaCl encryption key to be used by the
	// responder to NaCl seal the response
	PublicEncryptionKey [32]byte `msgpack:"publicEncryptionKey"`
	ChallengeData       []byte   `msgpack:"challengeData"`
	RequestData         []byte   `msgpack:"requestData"`
	Timestamp           int64    `msgpack:"timestamp"`
	ChallengeId         []byte   `msgpack:"challengeId"`
}

func Generate(signer crypto.Signer, challengeId []byte, challengeData []byte, requestData []byte) ([]byte, *[32]byte, error) {
	pubEncKey, privEncKey, err := box.GenerateKey(rand.Reader)
	if err != nil {
		return nil, nil, fmt.Errorf("generating encryption keys: %w", err)
	}

	inner, err := msgpack.Marshal(InnerChallenge{
		PublicEncryptionKey: *pubEncKey,
		ChallengeData:       challengeData,
		Timestamp:           time.Now().Unix(),
		ChallengeId:         challengeId,
		RequestData:         requestData,
	})

	if err != nil {
		return nil, nil, fmt.Errorf("marshaling inner challenge box: %w", err)
	}

	signature, err := echelper.SignWithTimeout(signer, inner, signingTimeoutDuration, signingTimeoutInterval)
	if err != nil {
		return nil, nil, fmt.Errorf("signing challenge: %w", err)
	}

	outerChallenge := &OuterChallenge{
		Sig: signature,
		Msg: inner,
	}

	outerChallengeBytes, err := outerChallenge.Marshal()
	if err != nil {
		return nil, nil, fmt.Errorf("marshaling outer challenge: %w", err)
	}

	return outerChallengeBytes, privEncKey, nil
}
